akka-cluster-sharding-typed/src/test/scala/docs/delivery/ShardingDocExample.scala

```diff
@@ -0,0 +1,223 @@
+
+    def apply(shardingProducerController: ActorRef[ShardingProducerController.Command[TodoList.Command]])
+        : Behavior[Command] = {
+      Behaviors.setup { context =>
+        new TodoService(context).start(shardingProducerController)
+      }
+    }
+
+  }
+
+  class TodoService(context: ActorContext[TodoService.Command]) {
+    import TodoService._
+
+    private implicit val askTimeout: Timeout = 5.seconds
+
+    private def start(
+        shardingProducerController: ActorRef[ShardingProducerController.Start[TodoList.Command]]): Behavior[Command] = {
+      val requestNextAdapter: ActorRef[ShardingProducerController.RequestNext[TodoList.Command]] =
+        context.messageAdapter(WrappedRequestNext.apply)
+      shardingProducerController ! ShardingProducerController.Start(requestNextAdapter)
+
+      Behaviors.receiveMessagePartial {
+        case WrappedRequestNext(next) =>
+          active(next)
+        case UpdateTodo(_, _, _, replyTo) =>
+          // not hooked up with shardingProducerController yet
+          replyTo ! Rejected
+          Behaviors.same
+      }
+    }
+
+    private def active(requestNext: ShardingProducerController.RequestNext[TodoList.Command]): Behavior[Command] = {
+      Behaviors.receiveMessage {
+        case WrappedRequestNext(next) =>
+          active(next)
+
+        case UpdateTodo(listId, item, completed, replyTo) =>
+          if (requestNext.bufferedForEntitiesWithoutDemand.getOrElse(listId, 0) >= 100)
+            replyTo ! Rejected
+          else {
+            val requestMsg = if (completed) TodoList.CompleteTask(item) else TodoList.AddTask(item)
+            context.ask[ShardingProducerController.MessageWithConfirmation[TodoList.Command], Done](
+              requestNext.askNextTo,
+              askReplyTo => ShardingProducerController.MessageWithConfirmation(listId, requestMsg, askReplyTo)) {
+              case Success(Done) => Confirmed(replyTo)
+              case Failure(_)    => TimedOut(replyTo)
+            }
+          }
+          Behaviors.same
+
+        case Confirmed(originalReplyTo) =>
+          originalReplyTo ! Accepted
+          Behaviors.same
+
+        case TimedOut(originalReplyTo) =>
+          originalReplyTo ! Rejected
+          Behaviors.same
+      }
+    }
+
+  }
+  //#producer
+
+  def illustrateInit(): Unit = {
+    Behaviors.setup[Nothing] { context =>
+      //#init
+      import akka.cluster.sharding.typed.scaladsl.ClusterSharding
+      import akka.cluster.sharding.typed.scaladsl.Entity
+      import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
+      import akka.cluster.typed.Cluster
+
+      val db: DB = ???
+
+      val system = context.system
+
+      val TypeKey = EntityTypeKey[ConsumerController.SequencedMessage[TodoList.Command]]("todo")
+
+      val region = ClusterSharding(system).init(Entity(TypeKey)(entityContext =>
+        ShardingConsumerController(start => TodoList(entityContext.entityId, db, start))))
```

Comment:
That is `ShardingConsumerController[TodoList.Command, TodoList.Command]`. The `B` type doesn't have to be the Command type of the Behavior, i.e. in this example AddTask and Complete task doesn't have to extend TodoList.Command.

I'm pondering if that flexibility is just utterly confusing and we should only have one type? wdyt @chbatey and @johanandren ?

Reply:
I'm leaning towards one type, but maybe I'm missing out on something?